### PR TITLE
Fako/buas projects

### PR DESCRIPTION
**Pull request by fako**

### CODE REVIEW

File: harvester/projects/fixtures/resources/buas-test.json [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+                "content-encoding": "gzip",
+                "x-xss-protection": "1; mode=block",
+                "transfer-encoding": "chunked",
+                "x-content-type-options": "nosniff",
+                "content-security-policy": "default-src 'self' *.elsevier.com cdn.plu.mx api.plu.mx d39af2mgp1pqhg.cloudfront.net ajax.googleapis.com *.pendo.io pendo-static-5644800010354688.storage.googleapis.com pendo-io-static.storage.googleapis.com *.adobedtm.com *.demdex.net cm.everesttech.net *.youtube.com i.ytimg.com *.vimeo.com *.altmetric.com d1bxh8uas1mnw7.cloudfront.net *.googletagmanager.com *.google-analytics.com *.hotjar.com cdn.cookielaw.org maps.googleapis.com www.google.com *.gstatic.com fonts.googleapis.com stats.g.doubleclick.net hcaptcha.com *.hcaptcha.com geolocation.onetrust.com *.onetrust.com https://js-agent.newrelic.com https://bam.nr-data.net https://bam-cell.nr-data.net 'unsafe-inline' 'unsafe-eval' data:; img-src * data: maps.gstatic.com *.googleapis.com;",
+                "strict-transport-security": "max-age=63072000; includeSubDomains, max-age=31536000 ; includeSubDomains"
+            },
+            "body": "buas/buas-projects.01.pii.json"
```
Comment (fako) [inline harvester/projects/fixtures/resources/buas-test.json:53]: Dit is een fixture file, die database rows aanmaakt tijdens test runs (zie Django docs). 

Het bijzondere van deze line is dat de body van de API response wordt ingeladen op basis van data die zich in S3 bevindt. Deze data bevat PII en mag daarom niet in de repo. Echter de wrapper data die te maken heeft met de response en welke request is gemaakt. Die mag wel in de repo.

Het is de ResourceFixturesSeedingTestCase die dit soort bijzondere fixtures in leest en het pad in de body property vervangt met de echte data van S3. Op deze manier de test data in de repo opslaan is relatief nieuw. De buas/buas-projects.01.pii.json file bevat 100 producten. Dat is heel fijn om de variatie van de data goed te kunnen waarnemen tijdens tests. Het is echter ondoenlijk om dat allemaal te anonimiseren. Dit is de gekozen tussenoplossing.

File: harvester/projects/models/resources/buas.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+from django.conf import settings
+
+from datagrowth.resources import HttpResource
+
+
+logger = logging.getLogger("harvester")
+
+
+class BuasPureProjectResource(HttpResource):
```
Comment (fako) [inline harvester/projects/models/resources/buas.py:11]: Deze class declareert hoe het systeem moet omgaan met de BUAS API. Alle API responses worden opgeslagen in de database, zodat bugs makkelijker te debuggen zijn. Bovendien doet herhaaldelijk "harvesten" dus niet altijd echte requesten. Waar mogelijk en waar toegestaan zal het systeem API responses herbruiken. Dit loopt allemaal via de HttpResource interface.

File: harvester/projects/sources/buas.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+            })
+        return persons
+
+    @classmethod
+    def get_owners(cls, node):
+        persons = cls.get_persons(node)
+        return [persons[0]] if persons else []
+
+
+OBJECTIVE = {
```
Comment (fako) [inline harvester/projects/sources/buas.py:65]: Dit is hoe een project eruit ziet qua data. In dit geval voor BUAS.

File: harvester/projects/tests/seeding/test_buas.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+from projects.models import Set, BuasPureProjectResource
+from projects.sources.buas import SEEDING_PHASES
+
+
+class TestBuasProjectSeeding(seeding.ResourceFixturesSeedingTestCase):
+
+    fixtures_directory = Path(settings.BASE_DIR, "projects", "fixtures", "resources", "buas")
+    resource_fixtures = ["buas-test"]
+    delta_fixtures = {
+        (BuasPureProjectResource, 1): ("body", "buas-projects.02.pii.json")
```
Comment (fako) [inline harvester/projects/tests/seeding/test_buas.py:20]: Hier geven we aan dat tijdens deltas (de niet initiele harvest, die waarschijnlijk een subset van de data is) een andere PII data file gebruikt moet worden. Hiermee simuleren we creates, updates en deletes voor in dit geval BUAS, maar elk bronsysteem heeft daarin natuurlijk eigen mechanismen.

File: harvester/projects/tests/seeding/test_buas.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+    }
+
+    entity = "projects"
+    source = "buas"
+    delete_policy = DeletePolicies.NO
```
Comment (fako) [inline harvester/projects/tests/seeding/test_buas.py:25]: Geen deletes ondersteund voor BUAS. Er zijn wel meer bronnen met deze policy.

File: harvester/projects/tests/seeding/test_buas.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+from datagrowth.resources.testing import ResourceFixturesMixin
+
+from core.constants import DeletePolicies
+from core.processors import HttpSeedingProcessor
+from testing.cases import seeding
+from projects.models import Set, BuasPureProjectResource
+from projects.sources.buas import SEEDING_PHASES
+
+
+class TestBuasProjectSeeding(seeding.ResourceFixturesSeedingTestCase):
```
Comment (fako) [inline harvester/projects/tests/seeding/test_buas.py:15]: Deze TestCase test puur het connecten met de bron en dat data correct wordt geupdate. Het richt zich op de HOE van de verbinding.

File: harvester/projects/tests/seeding/test_buas.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+            "because most data didn't come in through the delta"
+        )
+        new_title = "K3K Consortium development zero-emissions tourism mobility"
+        self.assertEqual(
+            self.set.documents.filter(properties__title=new_title).count(), 1,
+            "Expected title to get updated during delta harvest"
+        )
+
+
+class TestBUASProjectsExtraction(ResourceFixturesMixin, TestCase):
```
Comment (fako) [inline harvester/projects/tests/seeding/test_buas.py:57]: Deze TestCase haalt data op (uit de cache) tijdens de setup fase. Het testen richt zich op WAT er geharvest wordt. Klopt de data in alle velden? We proberen edge cases in de artificiele API responses op te nemen, zodat we zeker weten dat de cases gedekt zijn.

File: harvester/testing/cases/seeding.py [hunk 8]
```diff
@@ -126,3 +100,72 @@ def test_delta_seeding(self, become_processing_ids) -> list[HarvestDocument]:
                     self.assertTrue(doc.finished_at)
                 documents.append(doc)
         return documents
+
+
+class FactorySeedingTestCase(SeedingTestCase):
```
Comment (fako) [inline harvester/testing/cases/seeding.py:105]: Dit is een oudere manier van testen waarbij er een Factory is, die nep API responses maakt en daarna de data uit PII files haalt. Dit werkt minder makkelijk dan een fixture met de request info erin, omdat er in de achtergrond nog een hoop magie gebeurd met die factories. Het gebruikt factory_boy een zeer krachtige mock oplossing, die vooral nuttig is bij diep geneste datastructuren waarbij je niet alles helemaal realistisch wilt mocken voor elke TestCase. De HttpResource mocks zijn echter zo simpel dat de factory_boy Factories overkill zijn en eerder in de weg zitten dan helpen.

**Review by froedevrolijk — APPROVED** [review]
